File: webapp-vercel/lib/performance.ts
import { sampleStdev } from './indicators';

export interface PerformanceReport {
  starting_equity: number;
  ending_equity: number;
  total_return_pct: number;
  max_drawdown_pct: number;
  win_rate: number | null;
  sharpe_ratio: number | null;
  sortino_ratio: number | null;
  num_closed_trades: number;
}

export function computePerformance(equityCurve: number[], tradePnls: number[]): PerformanceReport {
  if (!equityCurve.length) throw new Error('equityCurve must have at least one point');
  const starting = equityCurve[0];
  const ending = equityCurve[equityCurve.length - 1];
  const totalReturnPct = starting ? ending / starting - 1 : 0;
  let peak = equityCurve[0], maxDrawdown = 0;
  for (const e of equityCurve) {
    peak = Math.max(peak, e);
    if (peak > 0) maxDrawdown = Math.max(maxDrawdown, (peak - e) / peak);
  }
  const periodReturns: number[] = [];
  for (let i = 1; i < equityCurve.length; i++) if (equityCurve[i - 1] !== 0) periodReturns.push(equityCurve[i] / equityCurve[i - 1] - 1);
  let winRate: number | null = null;
  if (tradePnls.length) winRate = tradePnls.filter((p) => p > 0).length / tradePnls.length;

  function sharpe(returns: number[]): number | null {
    if (returns.length < 2) return null;
    const sd = sampleStdev(returns);
    if (sd === 0) return null;
    const mean = returns.reduce((a, b) => a + b, 0) / returns.length;
    return (mean / sd) * Math.sqrt(returns.length);
  }
  function sortino(returns: number[]): number | null {
    if (returns.length < 2) return null;
    const downside = returns.map((r) => Math.min(0, r));
    const dsd = Math.sqrt(downside.reduce((s, d) => s + d * d, 0) / (downside.length - 1));
    if (dsd === 0) return null;
    const mean = returns.reduce((a, b) => a + b, 0) / returns.length;
    return (mean / dsd) * Math.sqrt(returns.length);
  }

  return {
    starting_equity: starting, ending_equity: ending, total_return_pct: totalReturnPct, max_drawdown_pct: maxDrawdown,
    win_rate: winRate, sharpe_ratio: sharpe(periodReturns), sortino_ratio: sortino(periodReturns), num_closed_trades: tradePnls.length,
  };
}

File: trading_agent/engine/performance.py
"""Performance metrics computed from an equity curve and closed-trade PnLs.

Purely descriptive statistics about what already happened in a run
(live `watch` session or `backtest` replay) — nothing here forecasts or
promises anything about the future. Sharpe/Sortino here are *per-tick*
ratios assuming a zero risk-free rate, not annualized figures comparable
to a fund's reported Sharpe ratio; they're a relative diagnostic for
"did this run's return look steady or lumpy," not a rigorous industry
metric.
"""
from __future__ import annotations

import math
from dataclasses import dataclass


@dataclass
class PerformanceReport:
    starting_equity: float
    ending_equity: float
    total_return_pct: float
    max_drawdown_pct: float
    win_rate: float | None
    sharpe_ratio: float | None
    sortino_ratio: float | None
    num_closed_trades: int


def compute_performance(equity_curve: list[float], trade_pnls: list[float]) -> PerformanceReport:
    if not equity_curve:
        raise ValueError("equity_curve must have at least one point")

    starting = equity_curve[0]
    ending = equity_curve[-1]
    total_return_pct = (ending / starting - 1) if starting else 0.0

    peak = equity_curve[0]
    max_drawdown = 0.0
    for e in equity_curve:
        peak = max(peak, e)
        if peak > 0:
            max_drawdown = max(max_drawdown, (peak - e) / peak)

    period_returns = [
        equity_curve[i] / equity_curve[i - 1] - 1
        for i in range(1, len(equity_curve))
        if equity_curve[i - 1] != 0
    ]

    win_rate = None
    if trade_pnls:
        win_rate = sum(1 for p in trade_pnls if p > 0) / len(trade_pnls)

    return PerformanceReport(
        starting_equity=starting,
        ending_equity=ending,
        total_return_pct=total_return_pct,
        max_drawdown_pct=max_drawdown,
        win_rate=win_rate,
        sharpe_ratio=_sharpe(period_returns),
        sortino_ratio=_sortino(period_returns),
        num_closed_trades=len(trade_pnls),
    )


def _stdev(values: list[float]) -> float:
    mean = sum(values) / len(values)
    return math.sqrt(sum((v - mean) ** 2 for v in values) / (len(values) - 1))


def _sharpe(returns: list[float]) -> float | None:
    if len(returns) < 2:
        return None
    stdev = _stdev(returns)
    if stdev == 0:
        return None
    mean = sum(returns) / len(returns)
    return (mean / stdev) * math.sqrt(len(returns))


def _sortino(returns: list[float]) -> float | None:
    if len(returns) < 2:
        return None
    downside = [min(0.0, r) for r in returns]
    downside_dev = math.sqrt(sum(d**2 for d in downside) / (len(downside) - 1))
    if downside_dev == 0:
        return None
    mean = sum(returns) / len(returns)
    return (mean / downside_dev) * math.sqrt(len(returns))
